Compiling error: function definition is marked dllimport

Hi,

When fastcdr is compiled and installed, I continue to build fastrtps, (compile on Windows and set `-DBUILD_SHARED_LIBS=ON`),

then I got lots of similar errors:

```sh
C:/ros/jazzy/install/include/fastcdr/Cdr.h:2792:21: error: function 'eprosima::fastcdr::Cdr& eprosima::fastcdr::Cdr::deserialize_type(eprosima::fastcdr::EncodingAlgorithmFlag, std::function<bool(eprosima::fastcdr::Cdr&, const eprosima::fastcdr::MemberId&)>)' definition is marked dllimport
 2792 |     Cdr_DllAPI Cdr& deserialize_type(
      |                     ^~~~~~~~~~~~~~~~
```
the reason is that when compiled as shared lib (dll on windows), `Cdr_DllAPI` is defined as `__declspec( dllexport )` when compiling fastcdr and `__declspec( dllimport )` when compiling fastrtps and includes fastcdr as a shared lib.


Then the included header file (for fastftps), is like this
```c++
//Cdr.h
    Cdr_DllAPI Cdr& deserialize_type(
            EncodingAlgorithmFlag type_encoding,
            std::function<bool (Cdr&, const MemberId&)> functor)
    {
        return (this->*deserialize_type_)(type_encoding, functor);
    }
```
the error message says that the function is marked as `dllimport` but it has a function body (definition) at the same time, which is an error.

The suggested change is to move the function body/definition from .h file to .cpp file.

I'd like to create a PR if the suggestion above is acceptable.

--- a/include/fastcdr/Cdr.h
+++ b/include/fastcdr/Cdr.h
@@ -530,10 +530,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            const uint8_t& octet_t)
-    {
-        return serialize(static_cast<char>(octet_t));
-    }
+            const uint8_t& octet_t);
 
     /*!
      * @brief This function serializes a character.
@@ -551,10 +548,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            const int8_t int8)
-    {
-        return serialize(static_cast<char>(int8));
-    }
+            const int8_t int8);
 
     /*!
      * @brief This function serializes an unsigned short.
@@ -563,10 +557,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            const uint16_t ushort_t)
-    {
-        return serialize(static_cast<int16_t>(ushort_t));
-    }
+            const uint16_t ushort_t);
 
     /*!
      * @brief This function serializes a short.
@@ -584,10 +575,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            const uint32_t ulong_t)
-    {
-        return serialize(static_cast<int32_t>(ulong_t));
-    }
+            const uint32_t ulong_t);
 
     /*!
      * @brief This function serializes a long.
@@ -605,10 +593,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            const wchar_t wchar)
-    {
-        return serialize(static_cast<uint16_t>(wchar));
-    }
+            const wchar_t wchar);
 
     /*!
      * @brief This function serializes an unsigned long long.
@@ -617,10 +602,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            const uint64_t ulonglong_t)
-    {
-        return serialize(static_cast<int64_t>(ulonglong_t));
-    }
+            const uint64_t ulonglong_t);
 
     /*!
      * @brief This function serializes a long long.
@@ -675,10 +657,7 @@ public:
      * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& serialize(
-            char* string_t)
-    {
-        return serialize(static_cast<const char*>(string_t));
-    }
+            char* string_t);
 
     /*!
      * @brief This function serializes a string.
@@ -2765,10 +2744,7 @@ public:
      */
     Cdr_DllAPI Cdr& begin_serialize_type(
             Cdr::state& current_state,
-            EncodingAlgorithmFlag type_encoding)
-    {
-        return (this->*begin_serialize_type_)(current_state, type_encoding);
-    }
+            EncodingAlgorithmFlag type_encoding);
 
     /*!
      * @brief Tells to the encoder the encoding of the type finishes.
@@ -2778,10 +2754,7 @@ public:
      * position that exceeds the internal memory size.
      */
     Cdr_DllAPI Cdr& end_serialize_type(
-            Cdr::state& current_state)
-    {
-        return (this->*end_serialize_type_)(current_state);
-    }
+            Cdr::state& current_state);
 
     /*!
      * @brief Tells to the encoder a new type and its members starts to be decoded.
@@ -2793,10 +2766,7 @@ public:
      */
     Cdr_DllAPI Cdr& deserialize_type(
             EncodingAlgorithmFlag type_encoding,
-            std::function<bool (Cdr&, const MemberId&)> functor)
-    {
-        return (this->*deserialize_type_)(type_encoding, functor);
-    }
+            std::function<bool (Cdr&, const MemberId&)> functor);
 
     /*!
      * @brief Encodes an optional in the buffer.
@@ -2850,16 +2820,7 @@ public:
      * encoded.
      */
     Cdr_DllAPI Cdr& operator <<(
-            const MemberId& member_id)
-    {
-        if (next_member_id_ != MEMBER_ID_INVALID)
-        {
-            throw exception::BadParamException("Member id already set and not encoded");
-        }
-
-        next_member_id_ = member_id;
-        return *this;
-    }
+            const MemberId& member_id);
 
     /*!
      * @brief Decodes an optional from the buffer.

--- a/src/cpp/Cdr.cpp
+++ b/src/cpp/Cdr.cpp
@@ -467,6 +467,12 @@ bool Cdr::resize(
 }
 
 Cdr& Cdr::serialize(
+        const uint8_t& octet_t)
+{
+    return serialize(static_cast<char>(octet_t));
+}
+
+Cdr& Cdr::serialize(
         const char char_t)
 {
     if (((end_ - offset_) >= sizeof(char_t)) || resize(sizeof(char_t)))
@@ -479,6 +485,18 @@ Cdr& Cdr::serialize(
     }
 
     throw NotEnoughMemoryException(NotEnoughMemoryException::NOT_ENOUGH_MEMORY_MESSAGE_DEFAULT);
+}
+
+Cdr& Cdr::serialize(
+        const int8_t int8)
+{
+    return serialize(static_cast<char>(int8));
+}
+
+Cdr& Cdr::serialize(
+        const uint16_t ushort_t)
+{
+    return serialize(static_cast<int16_t>(ushort_t));
 }
 
 Cdr& Cdr::serialize(
@@ -515,6 +533,12 @@ Cdr& Cdr::serialize(
 }
 
 Cdr& Cdr::serialize(
+        const uint32_t ulong_t)
+{
+    return serialize(static_cast<int32_t>(ulong_t));
+}
+
+Cdr& Cdr::serialize(
         const int32_t long_t)
 {
     size_t align = alignment(sizeof(long_t));
@@ -547,6 +571,18 @@ Cdr& Cdr::serialize(
     }
 
     throw NotEnoughMemoryException(NotEnoughMemoryException::NOT_ENOUGH_MEMORY_MESSAGE_DEFAULT);
+}
+
+Cdr& Cdr::serialize(
+        const wchar_t wchar)
+{
+    return serialize(static_cast<uint16_t>(wchar));
+}
+
+Cdr& Cdr::serialize(
+        const uint64_t ulonglong_t)
+{
+    return serialize(static_cast<int64_t>(ulonglong_t));
 }
 
 Cdr& Cdr::serialize(
@@ -772,6 +808,12 @@ Cdr& Cdr::serialize(
     }
 
     throw NotEnoughMemoryException(NotEnoughMemoryException::NOT_ENOUGH_MEMORY_MESSAGE_DEFAULT);
+}
+
+Cdr& Cdr::serialize(
+        char* string_t)
+{
+    return serialize(static_cast<const char*>(string_t));
 }
 
 Cdr& Cdr::serialize(
@@ -2150,6 +2192,40 @@ Cdr& Cdr::deserialize_array(
 
     throw NotEnoughMemoryException(NotEnoughMemoryException::NOT_ENOUGH_MEMORY_MESSAGE_DEFAULT);
 }
+
+Cdr& Cdr::begin_serialize_type(
+        Cdr::state& current_state,
+        EncodingAlgorithmFlag type_encoding)
+{
+    return (this->*begin_serialize_type_)(current_state, type_encoding);
+}
+
+
+Cdr& Cdr::end_serialize_type(
+        Cdr::state& current_state)
+{
+    return (this->*end_serialize_type_)(current_state);
+}
+
+Cdr& Cdr::deserialize_type(
+        EncodingAlgorithmFlag type_encoding,
+        std::function<bool (Cdr&, const MemberId&)> functor)
+{
+    return (this->*deserialize_type_)(type_encoding, functor);
+}
+
+Cdr& Cdr::operator <<(
+        const MemberId& member_id)
+{
+    if (next_member_id_ != MEMBER_ID_INVALID)
+    {
+        throw exception::BadParamException("Member id already set and not encoded");
+    }
+
+    next_member_id_ = member_id;
+    return *this;
+}
+
 
 Cdr& Cdr::serialize_bool_array(
         const std::vector<bool>& vector_t)

--- a/src/cpp/Cdr.cpp
+++ b/src/cpp/Cdr.cpp
@@ -2225,7 +2225,6 @@ Cdr& Cdr::operator <<(
     return *this;
 }
 
-
 Cdr& Cdr::serialize_bool_array(
         const std::vector<bool>& vector_t)
 {

--- a/src/cpp/Cdr.cpp
+++ b/src/cpp/Cdr.cpp
@@ -2200,7 +2200,6 @@ Cdr& Cdr::begin_serialize_type(
     return (this->*begin_serialize_type_)(current_state, type_encoding);
 }
 
-
 Cdr& Cdr::end_serialize_type(
         Cdr::state& current_state)
 {

